src/compute/src/arrangement/manager.rs

```diff
@@ -170,6 +178,128 @@ impl SpecializedTraceHandle {
             }
         }
     }
+
+    /// Turns this trace into a padded version that reports empty data for all times less than the
+    /// trace's current logical compaction frontier.
+    fn into_padded(self) -> Self {
+        match self {
+            Self::RowRow(trace) => Self::RowRow(trace.into_padded()),
+        }
+    }
+}
+
+/// Handle to a trace that can be padded.
+///
+/// A padded trace contains empty data for all times greater than or equal to its `padded_since`
+/// and less than the logical compaction frontier of the inner `trace`.
+#[derive(Clone)]
+pub struct PaddedTrace<Tr> {
+    /// The wrapped trace.
+    trace: Tr,
+    /// The frontier from which the trace is padded, or `None` if it is not padded.
+    padded_since: Option<Antichain<Timestamp>>,
```

Comment:
Please make this generic over `Tr::Time`.

Reply:
The `PaddedTrace` is currently specialized to `mz_repr::Timestamp` because that makes its `TraceReader` implementation straightforward since the compaction frontier is totally ordered, so it is either < or >= the inner trace's frontier. Considering generic timestamps would require us to consider partially ordered times which makes things more complicated. 

It's of course easy to make only the definition of `PaddedTrace` generic over the timestamp and leave the `TraceReader` impl specialized to `mz_repr::Timestamp`. But how would that improve things?

Reply:
Alright, if it makes the implementation substantially more complex, I'm good with keeping the specific type here. Would it make sense to restrict `Tr: TraceReader<Time=Timestamp>`? Alternatively, we could use the `TotalOrder` marker trait to guarantee total orderedness, but up to you!

Reply:
(In the past, we'd use `mz_repr::Timestamp` for the concrete type, and import `timely::progress::Timestamp`.)